src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,338 @@
+        //    do
+        //    {
+        //        listObjects.Add(streamReaderConverter.Read(ref this));
+        //        //At this point we're looking at the EndObject token for the object, need to advance.
+        //        Read();
+        //    }
+        //    while (TokenType != JsonTokenType.EndObject);
+        //    return listObjects;
+        //}
+
+        //internal void ReadArrayOfObjects<T1, T2>(IList<T2> objectList, Utf8JsonStreamReaderConverter<T1> streamReaderConverter) where T1 : T2
+        //{
+        //    if (objectList is null)
+        //    {
+        //        return;
+        //    }
+
+        //    if (Read() && TokenType == JsonTokenType.StartArray)
+        //    {
+        //        while (Read() && TokenType != JsonTokenType.EndArray)
+        //        {
+        //            var convertedObject = streamReaderConverter.Read(ref this);
+        //            if (convertedObject != null)
+        //            {
+        //                objectList.Add(convertedObject);
+        //            }
+        //        }
+        //    }
+        //}
+
+        internal string ReadNextTokenAsString()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            if (Read())
+            {
+                return _reader.ReadTokenAsString();
+            }
+
+            return null;
+        }
+
+        internal string GetCurrentBufferAsString()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            return Encoding.UTF8.GetString(_buffer);
+        }
+
+        internal IList<string> ReadStringArrayAsIList(IList<string> strings = null)
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            if (TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    string value = _reader.ReadTokenAsString();
+
+                    strings = strings ?? new List<string>();
+
+                    strings.Add(value);
+                }
+            }
+
+            return strings;
+        }
+
+        internal List<string> ReadStringArrayAsList(List<string> strings = null)
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            if (TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    string value = _reader.ReadTokenAsString();
+
+                    strings = strings ?? new List<string>();
```

Comment:
https://github.com/NuGet/NuGet.Client/pull/5529/files#r1414989994

> I would love better default capacity on every single List creation based on most likely end result to reduce allocations due to resizes and reduce memory usage. - @davkean 

> I like that idea. Since there are many, would you do that inline or would a static function be better to do this in a single place? - @donnie-msft 

> @davkean This might be interesting for that: https://github.com/NuGet/Home/pull/12124/files#diff-0e899c815846f62b058281b85f22b98df554f592455dcd732e44c72205ab7cc8R246.
I'm sure @zivkan would love to know what you think. - @nkolev92

Reply:
I like that solution @nkolev92 otherwise if feels like we're making a guess for what default to use.

Reply:
yeah, I think the idea would be to make a good guess, but I'm not sure how much better that'd be.

Reply:
I think 4 vs 0 has the same amount of allocations. 

tbh, I'd just table that for later.

Reply:
0 only causes the `List<T>` object to be allocated, where as 4 causes both the `List<T>` object and the backing array to be allocated. This is more allocations than needed for lists that never get any values.

IMO the fundamental problem is that this helper function is used everywhere, for all sorts of parts of the assets file schema, that have very different defaults.  I don't think it's possible to have a sane default here. Instead, the default would need to be passed down from the deserialization code, so can be adjusted for each call's default.

I think we can also consider this a future improvement. Even using List's default, the benchmark shows a significant perf improvement.  We can then use a PerfView trace to see how often `List<T>.Resize` appears on the call stack, or even as a once-off debug, put a `Console.WriteLine(strings.Count)` to see how large the spread of different values are.

Reply:
mkay, gonna resolve this for now then.